internal/controller/serverclaim_controller.go

```diff
@@ -305,131 +305,110 @@ func (r *ServerClaimReconciler) claimServer(ctx context.Context, log logr.Logger
-	}
-	log.V(1).Info("Matching server found", "Server", server.Name)
+	log.V(1).Info("Ensured ObjectRef for Server", "Server", selectedServer.Name)
+	return selectedServer, nil
+}
 
-	err = r.ensureObjectRefForServer(ctx, log, claim, server)
-	if err != nil {
-		return nil, err
+func (r *ServerClaimReconciler) isUnderMaintenanceQueue(ctx context.Context, log logr.Logger, server *metalv1alpha1.Server) (bool, error) {
+	if server.Status.State == metalv1alpha1.ServerStateMaintenance || server.Spec.ServerMaintenanceRef != nil {
+		log.V(1).Info("Server in or entering Maintenance, Hence can not be claimed")
+		return true, nil
+	}
+	// check if the current available state is a temerary state between multiple Maintenances states.
+	// We do not want to claim server while it is undergoing series of Maintenance in sequence.
+	serverMaintenancesList := &metalv1alpha1.ServerMaintenanceList{}
+	if err := clientutils.ListAndFilter(ctx, r.Client, serverMaintenancesList, func(object client.Object) (bool, error) {
+		serverMaintenance := object.(*metalv1alpha1.ServerMaintenance)
+		return serverMaintenance.Spec.ServerRef.Name == server.Name, nil
+	}); err != nil {
+		return true, err
+	}
+	if len(serverMaintenancesList.Items) == 0 {
+		return false, nil
 	}
-	log.V(1).Info("Ensured ObjectRef for Server", "Server", server.Name)
-	return server, nil
+	log.V(1).Info("Server has ongoing Maintenances, Hence can not be claimed")
+	return true, nil
 }
 
-func (r *ServerClaimReconciler) claimServerByReference(ctx context.Context, log logr.Logger, claim *metalv1alpha1.ServerClaim) (*metalv1alpha1.Server, error) {
-	log.V(1).Info("Trying to claim server by reference")
-	server := &metalv1alpha1.Server{}
-	if err := r.Get(ctx, client.ObjectKey{Name: claim.Spec.ServerRef.Name}, server); err != nil {
-		return nil, err
-	}
+func (r *ServerClaimReconciler) isServerClaimable(ctx context.Context, log logr.Logger, server *metalv1alpha1.Server, claim *metalv1alpha1.ServerClaim) bool {
 	if claimRef := server.Spec.ServerClaimRef; claimRef != nil && claimRef.UID != claim.UID {
 		log.V(1).Info("Server claim ref UID does not match claim", "Server", server.Name, "ClaimUID", claimRef.UID)
-		return nil, nil
+		return false
 	}
 	if server.Status.State != metalv1alpha1.ServerStateAvailable {
 		log.V(1).Info("Server not in a claimable state", "Server", server.Name, "ServerState", server.Status.State)
-		return nil, nil
+		return false
 	}
 	if server.Status.PowerState != metalv1alpha1.ServerOffPowerState {
 		log.V(1).Info("Server is not powered off", "Server", server.Name, "PowerState", server.Status.PowerState)
-		return nil, nil
+		return false
 	}
-	if claim.Spec.ServerSelector == nil {
-		return server, nil
+	isUnderMaintenance, err := r.isUnderMaintenanceQueue(ctx, log, server)
+	// is undergoing maintenance and not in Reserved State, we should not claim this server
+	if server.Status.State != metalv1alpha1.ServerStateReserved && (err != nil || isUnderMaintenance) {
```

Comment:
`server.Status.State` is always `Available` here, due to the check `server.Status.State != metalv1alpha1.ServerStateAvailable` above.